lib/web/apiserver.go

```diff
@@ -4118,6 +4239,68 @@ func contextWithMFAResponseFromRequestHeader(ctx context.Context, requestHeader
 	return ctx, nil
 }
 
+type wsBearerToken struct {
+	Token string `json:"token"`
+}
+
+type wsStatus struct {
+	Type    string `json:"type"`
+	Status  string `json:"status"`
+	Message string `json:"message,omitempty"`
+}
+
+var wsIODeadline = time.Second * 4
+
+// AuthenticateRequest authenticates request using combination of a session cookie
+// and bearer token retrieved from a websocket
+func (h *Handler) AuthenticateRequestWS(w http.ResponseWriter, r *http.Request) (*SessionContext, *websocket.Conn, error) {
+	ctx, err := h.validateCookie(w, r)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin:     func(r *http.Request) bool { return true },
+	}
+
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		return nil, nil, trace.ConnectionProblem(err, "Error upgrading to websocket: %v", err)
+	}
+	if err := ws.SetReadDeadline(deadlineForInterval(wsIODeadline)); err != nil {
```

Comment:
It doesn't quite make sense to pass something called `wsIODeadline` into a function named `deadlineForInterval`. `deadlineForInterval` is built for taking in an "interval" and returning a "deadline". I don't know that "interval" really makes sense in this context, seems like simplest would just be

```suggestion
	if err := ws.SetReadDeadline(time.Now().Add(wsIODeadline)); err != nil {
```

(note that `deadlineForInterval` was multiplying `wsIODeadline` by 2, so you may want to adjust it if you go with this suggestion).